server/groups/keycloak.go

```diff
@@ -530,24 +507,189 @@ func (k *KeycloakClient) HandleSAMLLogin(c *plugin.Context, user *mmModel.User,
+					"group_id", groupID,
+					"error", err)
+			}
+		}
+	}
+
+	finalTeamsToLeave := make(map[string]bool)
+	// Loop over teams to leave and check if it's in the list of teams to join.
+	// If it's in the teams to join, we don't need to remove the team membership.
+	for teamID := range proposedTeamsToLeave {
+		if _, exists := finalTeamsToJoin[teamID]; !exists {
+			finalTeamsToLeave[teamID] = true
 		}
 	}
 
+	finalChannelsToLeave := make(map[string]bool)
+	// Loop over channels to leave and check if it's in the list of channels to join.
+	// If it's in the channels to join, we don't need to remove the channel membership.
+	for channelID := range proposedChannelsToLeave {
+		if _, exists := finalChannelsToJoin[channelID]; !exists {
+			finalChannelsToLeave[channelID] = true
+		}
+	}
+	k.removeUserFromChannels(finalChannelsToLeave, user)
+	k.removeUserFromTeams(finalTeamsToLeave, user)
+	k.addUserToTeams(finalTeamsToJoin, user)
+	k.addUserToChannels(finalChannelsToJoin, user)
+
 	return nil
 }
 
+func (k *KeycloakClient) removeUserFromTeams(teamsToLeave map[string]bool, user *mmModel.User) {
+	for teamID := range teamsToLeave {
+		team, err := k.PluginAPI.Team.Get(teamID)
+		if err != nil {
+			k.PluginAPI.Log.Error("Failed to get team",
+				"team_id", teamID,
+				"error", err)
+			continue
+		}
+
+		// Don't remove user from the team if it's not group constrained.
+		// Groups get associated with teams when a group is associated to a channel within the team. So we cannot tell if the user was added to the team directly or through the group.
+		if team.GroupConstrained == nil || (team.GroupConstrained != nil && !*team.GroupConstrained) {
+			continue
+		}
+		// Check if they are a member of the team
+		member, err := k.PluginAPI.Team.GetMember(teamID, user.Id)
+		if err != nil {
+			if strings.ToLower(err.Error()) == "not found" {
+				k.PluginAPI.Log.Debug("User has already left the team", "team_id", teamID, "user_id", user.Id)
+			} else {
+				k.PluginAPI.Log.Error("Failed to get team member",
+					"user_id", user.Id,
+					"team_id", teamID,
+					"error", err)
+			}
+			continue
+		}
+		if member != nil && member.DeleteAt == 0 {
+			k.PluginAPI.Log.Debug("Removing user from team", "team_id", teamID, "user_id", user.Id)
+			if err = k.PluginAPI.Team.DeleteMember(teamID, user.Id, ""); err != nil {
+				k.PluginAPI.Log.Error("Failed to remove user from team",
+					"user_id", user.Id,
+					"team_id", teamID,
+					"error", err)
+				continue
+			}
+		}
+	}
+}
+
+// addUserToTeams adds a user to all teams associated with a group
+func (k *KeycloakClient) addUserToTeams(teamsToJoin map[string]bool, user *mmModel.User) {
+	for teamID := range teamsToJoin {
+		// Check if they are a member of the team
+		member, err := k.PluginAPI.Team.GetMember(teamID, user.Id)
+		if err != nil {
+			if strings.ToLower(err.Error()) != "not found" {
+				k.PluginAPI.Log.Error("Failed to get team member",
```

Comment:
It would be helpful if this error indicated a failure to add user to team, with cause being the failure to get team member.